---
title: "NIST Micro RM ROA Template"
author: "Nathan D. Olson"
date: "March 20, 2015"
output:
  html_document: default
  pdf_document:
    latex_engine: xelatex
    toc: yes
  toc: yes
bibliography: micro_rm_references.bib
---
```{r, echo = FALSE, message = FALSE}
source("rm_metadata.R")
```

```{r, echo = FALSE, message=FALSE}
library(png)
library(grid)
library(peprr)
library(dplyr)
library(knitr)
```

```{r echo=FALSE, message=FALSE} 
peprDB <- dplyr::src_sqlite(db_path)
```

# Introduction
## General background
Increasingly, high stakes decisions impacting public health and safety are being made using microbial genomic sequencing data [@Tang2014]. For example, whole genome sequencing was recently used as part of an investigation of the European 2011 Escherichia coli O1O4:H4 sprout associated outbreak [@Grad2012]. As the stakes increase so does the required level of confidence in the measurement. This reference material as well as three other microbial genomic DNA RM8375-8378 were developed to help advance the measurement assurance of microbial genomic sequencing and DNA sequencing in general. The development of these reference material was supported through an Interagency agreement with the Food and Drug Administration's Center for Devices, for use in validating DNA sequencing platforms for clinical applications.

## Material Description
```{r echo = FALSE,child='rm_strain_selection_info.Rmd' }
## Add text specific to rm regarding strain selection and publically available reference genome sequence
```


Loftstrand Labs[^lofstrand] Limited grew a large batch of `r rm_strain` provided by `r strain_source`  to produce ~ `r (rm_vial_number * 3)/1000` mg of total extracted DNA, divided equally into `r rm_vial_number` vials (Appendix: [Section A][], [Section B][]). Each unit of `r rm_number` is approximately `r rm_mass` ug of extracted genomic DNA in Tris buffer.  The material is stored in 0.5 ml screw cap microcentrifuge tubes (Item number 1405-9710 USA Scientific [^usa]).

```{r echo = FALSE,child='rm_specific_culture_ext.Rmd' }
## Add text specific to rm regarding culturing conditions and DNA extraction.
```
 
This RM is isolated DNA rather than live cells because cells can mutate with each cell division, and the genome sequence may not be stable over time for live cells. Extracting DNA from a large batch of cells helps ensure that all vials contain essentially the same DNA sequences. Even though the DNA in the cells is likely to mutate at low frequency during the growth, the resulting mutations will be in extremely small proportions of the overall cells unless they confer a selective advantage and occur early in the culture process. Even if any mutations confer a selective advantage, the DNA from the large batch of cells was mixed before aliquoting to minimize the variability in the proportions of mutations among vials. 
 
## Intended Use
The intended use of `r rm_number` is to help assess performance of high-throughput DNA sequencing.  There are two primary intended use cases for the RM:  

1. DNA sequencing and data analysis method validation and quality control as part of a laboratory quality control program;     
2. microbial genome sequencing for DNA sequencing including genome assembly, read mapping based variant calling methods.   

The genomic DNA is intended to be analyzed in the same way as any other sample a laboratory would analyze extracted DNA, such as through the use of a genome assembly or variant calling bioinformatic pipeline (Fig. 1).  Because the RM is extracted DNA, it does not assess pre-analytical steps such as DNA extraction. It does however, challenge sequencing library preparation, sequencing machines, base calling algorithms, and the subsequent bioinformatic analysis.  This RM is not intended to assess bioinformatics steps such as strain identificaiton, phylogenetic analysis, or genome annotation. It is important to recognize the genome sequence is provided as an “Information Value” rather than “Certified Value” because we were not yet fully confident that we accounted for all known biases.  However, we evaluated the individual base calls and genome assembly for a number of known sources of bias, e.g. repeat regions, and included additional information regarding these sources of bias (see vcf and gff files). 

```{r echo = FALSE, fig.width = 6, fig.align='center', fig.cap= "Generic measurement process for genome assembly and variant calling intended use cases.  For sequencing quality control the results from the sequencing step would be compared to the reference material genome sequence to determine base level sequencing accuracy."}
grid.raster(readPNG("measurement_process.png"))
```
  

## Genome Sequence as a reference value
Characterizing a microbial genome RM at each of the $`r rm_genome_size/1000000`\times 10^6$ positions poses a significant metrological challenge. The consensus base call at each genomic position is a “nominal property”, because it is not a numerical value.  A previous sequence-based SRM 2374 was certified primarily based on capillary (or “Sanger”) sequencing, a relatively mature sequencing technology that is not easily scalable to whole genomes.  The maturity of capillary sequencing as well as the small size of SRM 2374 (< 100,000 bases) allowed for manual curation of capillary sequencing traces across the entire DNA sequences, and the assignment of qualitative (non-probabilistic) statements about the uncertainty of each base call.

To characterize `r rm_number`, we relied primarily on a variety of less mature “Next Generation Sequencing” (NGS) technologies. Unlike capillary sequencing, NGS technologies can sequence a whole genome at a reasonable cost.  NGS technologies measure many reads at each position, from which variant callers and consensus base callers calculate probabilistic estimates of uncertainty based on Bayesian statistics and assuming binomial sampling.  Unfortunately, these probabilistic estimates of uncertainty are rarely accurate due to various sources of bias and error in the sequencing and bioinformatics (Fig. 1).  Therefore, to develop a characterized reference genome sequence we first generated an assembly of the whole genome then performed a base level analysis of the genome assembly.  

The sources of bias and error associated with different steps in the sequence measurement process are indicated in the cause-effect diagram below (Fig. 2). Please see Olson (et al. _in prep_) for a detailed discussion of these sources of error.

```{r echo = FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Cause and effect diagram indicating the sources of bias and error for the individual components of the sequence measurement process."}
grid.raster(readPNG("MicroRMCauseEffect.png"))
```

In addition to the sources of error and bias associated with the measurement there is an unknown degree of biological variability within the batch.  The genomic DNA was homogeneized prior to aliquoting into the individual vials but the biologically variability within the population of cells from which the DNA was extracted is unknown.  The mutations that make up this variability include single nucleotide polymorphisms (SNPs), insertions and deletions (indels), as well as structural variants.  Low frequency SNPs and indels can be mistaken as sequencing errors and low frequency structural variants can lead to ambiguities in the genome assembly.  Our characterization methods did not attempt to identify low frequency biologically variability within the RM batch, however, when interpreting the reference material characterization procedure results the potential for these low frequency mutations should be considered.
    
Genomic purity, genome annotation, homopolymer length, and identification of repeat regions are provided as information values and have not been evaluated for sources of bias an error.  The genomic purity of the reference material was assessed using a taxonomic sequence classification algorithm, the accuracy of the method was evaluated using simulated sequence data (Olson et al. _in prep_).  However, sources of bias associated with the measurement, namely accuracy of the database used for taxonomic assignment and the assignment algorithm have not been investigated. Genome annotation is a major field of study within the biological scientific community and represents an unsolved problem in genomics, for well characterized strains such as `r rm_strain` the annotation has been highly curated, however the accuracy of the annotation is unknown. The accuracy of the homopolymer length and repeat region information is dependent on the reference genome sequence and therefore share the same sources of error and bias as the methods used to determine the reference genome sequence.

\pagebreak

# Characterization Process
## Sequence Characterization
### Sequencing Measurements

```{r echo = FALSE,child='rm_specific_nist_sequencing.Rmd' }
## Add text specific to rm regarding sequencing measurements made at NIST.
```
    
Eight vials from the `r rm_number` lot were randomly selected. In addition to NIST measurements, the material was sequenced using the Pacific Biosciences [^pacbio] RSII, and optical mapping was by OpGen through their MapIt service.  

### Sequence Analysis Methods
For the genome assembly we started with a _de novo_ genome assembly and validated the assembly using optical mapping and sequencing data generated using orthogonal measurement methods. Long read sequence data generated with the Pacific Biosciences [^pacbio] RSII generate sequence reads with a median length of around 8kb, which is long enough to spand the ribosomal operon repeat regions, often resulting in a complete genome assembly [@Koren2013]. While the long reads allow for the assembly of a complete genome the relatively high error rate ~11\%[^pb-error] (primarily insertions and deletions) for individual reads can lead to small variant errors. 

To validate the assembly orthogonal methods optical mapping and short read sequencing were used.  Optical mapping provides information regarding the overall structure of the genome based on the location of restriction sites within long fragments (average size > 200Mb).  Comparison of the fragment patterns between the optical mapping data and _in-silico_ generated fragment patterns from the genome assembly indicates potential misassemblies in the genome assembly [@Mendelowitz2014]. The Optical mapping data was generated by a commercial vendor OpGen[^opgen] and sources of bias and error have not been fully investigated. The due to the resolution limit of the optical mapping method it is unable to detect misassemblies smaller than 3 kb.  

Potential misassemblies smaller than 3 kb were identified using short read data generated using orthogonal sequencing technologies Illumina[^illumina] MiSeq, and Ion Torrent[^iontorrent] PGM and Pilon [@Walker2014], software developed to identify and correct misassemblies in microbial genomes.  The sources of bias associated with each of the orthogonal methods were not fully investigated and agreement between the methods supports the hypothesis that the assembly is correct. 

Disagreements between the genome assembly and the optical mapping or short read results were investigated for sources of bias to account for the discrepancy. When appropriate the genome assembly was corrected.  If we were unable to identify the source of the bias that caused the misassembly the genome sequence is annotated appropriately to indicate region where the assembly is ambiguous. It is important to note that agreement between orthogonal methods does not guarantee a correct result as they may all be susceptible to the same source(s) of bias. Particularly, we identify repetitive regions where multiple assembly methods may be most prone to errors. We intend to update the genome assembly as new data is collected and errors in the assembly are revealed.
    
Base level analysis was performed by evaluating the proportion of sequencing reads with bases supporting the genome assembly base call at individual genome positions, within the the lot (base purity) and among individual vials (base homogeneity). Short read sequencing data generated using the MiSeq and PGM were used for base level analysis. Base purity was defined for the two platform individually, using samtools mpileup [@Li2009]. Base homogeneity, was assessed using varscan somatic variant caller [@Koboldt2009], using the methods described in RM8398 report of analysis. Additional genome annotation information was obtained to provide context for base level sequence analysis include gene annotation, identification of repeat regions, and homopolymer length. Genome annotation was performed using the RAST server[^rast]. Repeat regions were identified using mumer [@Delcher2002].  

The purity of the reference material in terms of presence of DNA from sources other than `r rm_strain` was assessed using the metagenomic taxonomic read classification algorithm PathoScope [@Hong2014].  This method uses an expectation maximization algorithm where the sequence data are first mapped to a database comprised on all sequence data in the Genbank nt database, then through an iterative process re-assigns ambiguously mapped reads to based on the proportion of reads mapped unambigously to individual taxa in the database. 

#### Method Reproducibility
The methods used to characterize the reference material were developed in a manner to facilitate reuse and ensure computational reproducibility, obtaining the same results when starting with the same set of measurement data.  To achive this goal the data, computer code, as well as the computing environment were all made publically available.  The raw sequence data is archived at the Genbank Sequence Read Archive ([ncbi.nlm.nih.gov/sra](http://www.ncbi.nlm.nih.gov/sra)), see table 1 in the summary of sequencing data results section for accession numbers.  The code used to characterize the material is available through GitHub as a python bioinformatic program ([github pepr](http://www.github/nate-d-olson/pepr)) and an R package ([github peprr](http://www.github/nate-d-olson/peprr)).  The python bioinformatics pipeline performs the initial sequence analysis and uses a Docker environment for dependencies [docker.org](http://www.docker.com). A flowchat of the bioinformatic pipeline used to characterize the RM can be found in Appendix [Section D][]. The R package, includes functions for loading the output from the bioinformatics pipeline into an sqlite database and generating the summary figures and tables found within this report of analysis.  

\pagebreak

### Sequence Analysis Results
#### Summary of Sequencing Datasets
```{r echo = FALSE,child='rm_specific_dataset_summary.Rmd' }
## Add text specific to rm regarding general summary of sequencing datasets.
```

```{r echo = FALSE, message=FALSE, warning=FALSE}
kable(seq_summary_table(peprDB), 
      round = 0, row.names = FALSE, caption = "Summary of sequencing datasets")
```

\pagebreak

#### Genome Assembly
```{r echo = FALSE,child='rm_specific_assembly_summary.Rmd' }
## Add text specific to rm regarding general summary of genome assembly
```

```{r echo = FALSE,child='rm_specific_optical_map_summary.Rmd' }
## Add text specific to rm regarding general summary of optical mapping results
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Assembly of optical mapping data."}
grid.raster(readPNG("opgen_assembly.png"))
```

```{r echo = FALSE,child='rm_specific_assembly_validation_summary.Rmd' }
## Add text specific to rm regarding general summary of assembly validation results.
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap= "Comparison of optical map data to genome assembly."}
# __SPECIFIC TO RM__
grid.raster(readPNG("opgen_assembly_comparison.png"))
```

```{r echo =FALSE, message=FALSE, warning=FALSE}
## need a better way to present this information
#kable(pilon_changes_table(peprDB), 
#      row.names = FALSE,
#      caption = "Changes made to reference genome with Pilon")
```

\pagebreak

#### Base Level Analysis
##### Base Purity
```{r echo = FALSE,child='rm_specific_base_level_purity_summary.Rmd' }
## Add text specific to rm regarding general summary of base level purity.
```

```{r echo = FALSE, message = FALSE, warning=FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Comparison of base purity values for PGM and MiSeq."}
purity_scatter_plot(peprDB)
```

```{r echo =FALSE, message=FALSE, warning=FALSE}
## not sure how much this table really adds - excluding for now
# kable(low_purity_table(peprDB), 
#       row.names = FALSE,round = 2,
#       caption = "Positions with base level purity values less than 98% for both platforms.")
```

```{r echo = FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Bias metrics for low purity positions.", message=FALSE, warning=FALSE}
low_pur_metrics_figure(peprDB)
```

##### Sequence homogeneity
```{r echo = FALSE,child='rm_specific_base_level_homogeneity_summary.Rmd' }
## Add text specific to rm regarding general summary of base level homogeneity.
```

```{r echo =FALSE}
kable(homogeneity_sig_table(peprDB, rename_cols = TRUE), 
      row.names = FALSE,
      caption = "Pairwise variant analysis results")
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Among sample variation in purity values."}
homogeneity_point_line_figure(peprDB)
```

##### Genome Annotation
```{r echo = FALSE,child='rm_specific_genome_annotation_summary.Rmd' }
## Add text specific to rm regarding general summary of genome annotation.
```

#### Genomic Purity
```{r echo = FALSE,child='rm_specific_genomic_purity.Rmd' }
## Add text specific to rm regarding genomic purity.
```
```{r echo = FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Proportion of reads from contaminant DNA.", warning=FALSE, message=FALSE}
contam_counts_figure(peprDB, rm_genus)
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.height = 6, fig.cap= "Breakdown of contaminants by organism.",fig.retina=TRUE}
contam_point_line_figure(peprDB, rm_genus)
```

\pagebreak

## DNA Fragment Size Characterization
### DNA Fragment Size Stability and Homogeneity
`r rm_number` is homogeneous and stable under the recommended storage conditions, although there is evidence that DNA fragment size may change over time, as indicated by the confidence of reviewers.  Samples were exposed to aging and handling conditions outside those recommended for normal use in a controlled experimental setting.  The results of these experiments indicate the only variability in unit-to-unit stability may come after advanced aging.  No other observable differences were observed for handling or storage conditions.
The examination of several vials of material across the lot allocation by PFGE constitutes the homogeneity study for this material; no inhomogeneity is observed from unit-to-unit.

### Description of Homogeneity/Stability Study
Vials of RM material were selected at random to assess homogeneity and stability of the DNA.  For homogeneity measurements this included the first and last pooled aliquot as well as random vials from regular intervals across the lot.  For the stability study, a random number generator was used to determine the vials to measure.

An isochronous stability study was performed to ensure integrity of DNA fragment size for candidate `r rm_number`over a variety of storage and handling conditions.  Figure **XVZ** shows the design of the study, including the experimental treatments for each condition studied.  Two time courses and three handling conditionswere used to mimic accelerated aging, recommended storage conditions and normal laboratory processing.

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap= "Stability study experimental design."}
grid.raster(readPNG("stability.png"))
```

The eight-week time course was started first, followed by a two-week time course.  The study was conducted such that all time courses ended on the same day.  A -20°C temperature control was used since this is the recommended storage condition of candidate `r rm_number`.  Triplicate vials were measured at each time point and condition to ensure there was enough replication to determine a qualitative difference.

For handling conditions, triplicate vials were subjected to pipetting, vortexing and multiple freeze-thaw cycles.  For vials measuring the effects of pipetting, the entire volume of the tube (~50μL) was pipetted vigorously up and down 10 times.  The effects of vortexing were observed by vortexing the entire volume of the vial for 10 seconds.  To measure the effects of freeze/thaw cycles on DNA fragment size, one cycle of freeze/thaw was equivalent to the entire vial being placed at room temperature for one hour to completely thaw the liquid, then being placed back at -20°C overnight.  The study looked at the effects of 2 and 5 complete freeze/thaw cycles.

All temperature-based conditions were monitored with a data logger to track temperature, humidity and dew point over the entirety of the time-course.  All samples were stored in opaque boxes to minimize exposure the light.  At the conclusion of the time-course samples were placed in a -20°C freezer until measurements were made.

### Analysis of Homogeneity and Stability Study
To assess the homogeneity and stability of the DNA fragment size distribtion the proportion of DNA in a PFGE gel lane for each sample was determined for regions between ladder markers. These proportions were then compared using an ANOVA and a Tukey's HSD test was used to identify pairwise statistical differences. The gels were processed in a semi-automated manner were the parameters using the identify and label markers were manually set then used to process the gels in an automated manner. Prior to processing the gels were first manually cropped to remove regions of the gel to the left and right of the outer lanes and the gel wells used to load the gels. The gel wells were removed to eliminate potential biases due to presence of residual DNA in the wells. Finally the images were converted to grey scale prior to analysis. Code used to analyze the gels available at __%%TODO%%__.

#### Homogeneity Study Results
__%%TODO%%__ DNA fragment homogeneity results


#### Stability Study Results
```{r echo = FALSE, fig.align='center', fig.show = 'asis', fig.cap= "Comparison of DNA proportions for size ranges between control and treatment."}
#stability_tt_heatmap
```

\pagebreak

# Appendix

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap= "Certificate of analysis provided by the reference material manufacturer."}
# __SPECIFIC TO RM__
grid.raster(readPNG("Lofstrand_certificate_of_analysis.png"))
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.width=3, fig.height=3, fig.align='center', fig.cap= "RM tube label"}
# __SPECIFIC TO RM__
grid.raster(readPNG("RMLabel.png"))
```

```{r echo = FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap= "Bioinformatic pipeline flowchart."}
grid.raster(readPNG("bioinf-pipeline-flowchart.png"))
```

\pagebreak

## Section D
__%%TODO%%__ Example gel analysis

\pagebreak

# References

[^opgen]: OpGen Inc. [opgen.com](http://opgen.com) 708 Quince Orchard Road Gaithersburg, MD 20878 USA
[^usa]: USA Scientific, Inc. [usascientific.com](http://www.usascientific.com) PO Box 3565 Ocala, FL 34478 USA
[^illumina]: Illumina Inc., [illumina.com/](http://www.illumina.com/) 5200 Illumina Way San Diego, CA 92122 USA
[^iontorrent]: Life Technologies Corp., [iontorrent.com/](http://www.iontorrent.com/) 7000 Shoreline Court #201, South San Francisco, CA 94080 USA
[^pacbio]: Pacific Biosciences of California Inc. [pacificbiosciences.com/](http://www.pacificbiosciences.com/) 1380 Willow Rd. Menlo Park, CA 94025 USA
[^rast]: Rapid Annotaiton Using Subsystem Technology (RAST) server [rast.nmpdr.org](http://rast.nmpdr.org/rast.cgi)
[^lofstrand]: Lofstrand Labs Limited 7961 Cessna Avenue, Gaithersburg, MD 20879 [lofstrand.com/](http://www.lofstrand.com/)
[^pb-error]: Manufacturer stated error rate, in "Understanding Accuracy in SMRT Sequencing" [http://www.pacificbiosciences.com](http://www.pacificbiosciences.com/pdf/Perspective_UnderstandingAccuracySMRTSequencing.pdf)
